deps/ipc-util/ipc-util/pipe.h

```diff
@@ -33,16 +33,39 @@ typedef struct ipc_pipe_server ipc_pipe_server_t;
 typedef struct ipc_pipe_client ipc_pipe_client_t;
 
 typedef void (*ipc_pipe_read_t)(void *param, uint8_t *data, size_t size);
+typedef void (*ipc_pipe_connect_t)(void *param);
+typedef bool (*ipc_pipe_disconnect_t)(void *param);
 
 bool ipc_pipe_server_start(ipc_pipe_server_t *pipe, const char *name,
 			   ipc_pipe_read_t read_callback, void *param);
+void ipc_pipe_server_set_connect_callback(ipc_pipe_server_t *pipe,
+					  ipc_pipe_connect_t connect_callback,
+					  void *param);
+void ipc_pipe_server_set_disconnect_callback(
+	ipc_pipe_server_t *pipe, ipc_pipe_disconnect_t disconnect_callback,
+	void *param);
+void ipc_pipe_server_disconnect(ipc_pipe_server_t *pipe);
 void ipc_pipe_server_free(ipc_pipe_server_t *pipe);
+bool ipc_pipe_server_write(ipc_pipe_server_t *pipe, const void *data,
+			   size_t size);
+
+bool ipc_pipe_server_return_start(ipc_pipe_server_t *pipe, const char *name);
```

Comment:
I couldn't really determine a better way to name it. The only alternative I had in mind felt far worse (client_server and server_client). I can't just drop server and client from the return types, because how do you distinguish which side of the conversation the return is dealing with? Would love to hear alternative suggestions.

Reply:
I used `ipc_pipe_server_connect` (version of `ipc_pipe_server_start` that does not create the pipe), and `ipc_pipe_client_create` (`ipc_pipe_client_open` but it creates the pipe) in another project

Somewhat unhappy about the lack of symmetry, if we'd want to rename everything I'd maybe go with
- `ipc_pipe_server_create` + `ipc_pipe_server_connect`
- `ipc_pipe_client_connect` + `ipc_pipe_client_create`

`...create` seems like a better option for the client, since you're not really starting anything

Reply:
I don't think I agree that "server" <=> "always reads". A web server needs to send information back to the client that it must read, but you still refer to it as the server because it's the host that the connection is made to.

Reply:
Unfortunately, this is not a web server, and the limitations of named pipes mean that there's strong incentives for keeping reading confined to one side and writing confined to another; even better naming might be
- `ipc_pipe_reader_create` + `ipc_pipe_reader_connect`
- `ipc_pipe_writer_create` + `ipc_pipe_writer_connect`

What's your take on what server is supposed to mean in current code (before your modifications)?

Reply:
When I used server as part of the code, my interpretation of the term was "the host/owner of the pipe that something else is connecting to" and client to mean "the application that is connecting to a pipe hosted/owned by a server".

Reply:
Any thoughts then on how to discourage people from accidentally writing code that could deadlock?

Reply:
Well, that's why there's a difference between a server and server return, and client and client return. Server and client determines who is the host and who is the one connecting, while return determines which side is receiving the message.

Reply:
That doesn't sound like it would prevent anyone from using whichever function that takes a `ipc_pipe_server_t*`; helping people distinguish should imo involve something along the lines of `ipc_pipe_server_t*` + `ipc_pipe_server_return_t*`

Reply:
Ahh, I see now. (If I'm understanding you correctly) it's not about strict definitions of terms, but minimizing development errors. That pipes that send and pipes that receive can both have the type of "server" is a problem because when direction isn't implicit in the name, you can as a developer forget which pipe is for which direction, and that may lead to difficult to debug/unpredictable deadlocks.

I will rename the types using your reader/writer suggestion. I may also have to specifically make some kind of transaction pipe type for the initial connection which allows both a read and write, but only for that specific kind of communication. Having an abstracted API for that will probably also make it easier to port the library to POSIX systems down the line.